truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -264,77 +264,115 @@ public Object ulong2num(long num) {
-                    x = x | (a_byte << (j * 8));
-                }
-                longs[i] = x;
+
+            // If we're not giving a twos comp answer then a leading
+            // zero byte should be discarded.
+            if (!twosComp && bi_bytes[0] == 0) {
+                bi_length--;
             }
-            if (fraction != 0) {
-                long x = 0;
-                for (int i = 0; i < fraction; i++) {
-                    long a_byte = bytes[bytes.length - 1 - limit * 8 - i] & 0xffl;
-                    x = x | (a_byte << (i * 8));
+            int bytes_to_copy = Math.min(bi_length, num_bytes);
+            if (msw_first) {
+                // We must copy the LSBs if the buffer would overflow,
+                // so calculate an offset based on that.
+                int offset = bi_bytes.length - bytes_to_copy;
+
+                for (int i = 0; i < bytes_to_copy; i++) {
+                    bytes[i] = bi_bytes[offset + i];
+                }
+                if (negative) {
+                    for (int i = 0; i < num_bytes - bytes_to_copy; i++) {
+                        bytes[i] = -1;
+                    }
+                }
+            } else {
+                for (int i = 0; i < bytes_to_copy; i++) {
+                    bytes[i] = bi_bytes[bi_bytes.length - 1 - i];
                 }
                 if (negative) {
-                    for (int i = fraction; i < 8; i++) {
-                        x = x | (0xffl << (i * 8));
+                    for (int i = bytes_to_copy; i < num_bytes; i++) {
+                        bytes[i] = -1;
                     }
                 }
-                longs[limit] = x;
             }
-            if (negative) {
-                for (int i = blank_start; i < length; i++) {
-                    longs[i] = -1;
+
+            // Swap bytes around if they aren't in the right order for
+            // the requested endianness.
+            if (bigEndian ^ msw_first && word_length > 1) {
+                for (int i = 0; i < num_words; i++) {
+                    for (int j = 0; j < word_length / 2; j++) {
+                        int pos_a = i * word_length + j;
+                        int pos_b = (i + 1) * word_length - 1 - j;
+                        int a = bytes[pos_a];
+                        bytes[pos_a] = bytes[pos_b];
+                        bytes[pos_b] = a;
+                    }
                 }
             }
-            return ArrayHelpers.createArray(getContext(), longs, longs.length);
+            return ArrayHelpers.createArray(getContext(), bytes, bytes.length);
         }
+
+
     }
 
     @CoreMethod(names = "rb_absint_bit_length", isModuleFunction = true, required = 1)
-    public abstract static class BignumBitLengthNode extends CoreMethodArrayArgumentsNode {
+    public abstract static class BignumAbsBitLengthNode extends CoreMethodArrayArgumentsNode {
+        @Specialization(guards = "isRubyBignum(num)")
```

Comment:
Not sure, it's pretty small really. Compares the signum field with zero, and either returns self or calls `this.negate()`. That in turn calls a constructor passing in the existing byte array and the signum. Most real world uses are likely to simply return `this`